pxtlib/emitter/decompiler.ts

```diff
@@ -817,5 +852,70 @@ ${output}</xml>`;
                 }
             }
         }
+
+        function pushScope() {
+            scopes.push({})
+        }
+
+        function popScope() {
+            scopes.pop();
+        }
+
+        function addVariable(name: string) {
+            scopes[scopes.length - 1][name] = getNewName(name);
+        }
+
+        function addVariableDeclaration(node: VariableDeclaration): boolean {
+            if (node.name.kind !== SK.Identifier) {
+                error(node, Util.lf("Variable declarations may not use binding patterns"))
+                return false;
+            }
+            else if (!node.initializer) {
+                error(node, Util.lf("Variable declarations must have an initializer"))
+                return false;
+            }
+            const name = (node.name as ts.Identifier).text
+            addVariable(name)
+            return true;
+        }
+
+        function getVariableName(name: string) {
+            const existingName = findVariableName(name);
+            if (existingName) {
+                return existingName;
+            }
+            else {
+                addVariable(name);
+                return name;
+            }
+        }
+
+        function findVariableName(name: string) {
+            for (let i = scopes.length - 1; i >= 0; i--) {
+                if (scopes[i][name]) {
+                    return scopes[i][name];
+                }
+            }
+            return undefined;
+        }
+
+        function isScopeEnd(n: NextNode): n is "ScopeEnd" {
+            return typeof n === "string"
+        }
+
+        function getNewName(name: string) {
+            if (!takenNames[name]) {
+                takenNames[name] = true;
+                return name;
+            }
+
+            for (let i = 0; ; i++) {
```

Comment:
if the variable name is "i", we could be smart and try to pick "j" and so forth, then fallback to numbers.